source/common/pool_impl.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+#include "nighthawk/common/exception.h"
+#include "nighthawk/common/poolable.h"
+
+#include "common/common/assert.h"
+#include "common/poolable_impl.h"
+
+namespace Nighthawk {
+
+template <typename Poolable> class PoolImpl {
+public:
+  using PoolDeletionDelegate = std::function<void(Poolable*)>;
+  using PoolInstanceConstructionDelegate = std::function<Poolable*(void)>;
+  using PoolInstanceResetDelegate = std::function<void(Poolable&)>;
+  using PoolablePtr = std::unique_ptr<Poolable, PoolDeletionDelegate>;
+
+  PoolImpl(PoolInstanceConstructionDelegate&& construction_delegate = nullptr,
+           PoolInstanceResetDelegate&& reset_delegate = nullptr)
+      : construction_delegate_(construction_delegate), reset_delegate_(reset_delegate) {}
+
+  ~PoolImpl() {
+    while (!pool_.empty()) {
+      Poolable* poolable = pool_.back().get();
+      all_.erase(std::remove(all_.begin(), all_.end(), poolable), all_.end());
+      pool_.pop_back();
+    }
+    // Inform the in-flight poolables that they are own their own now.
+    for (auto poolable : all_) {
+      poolable->mark_orphaned();
+    }
+  }
+
+  void addPoolable(std::unique_ptr<Poolable>&& poolable) {
```

Comment:
I think the `&&` isn't needed; the `unique_ptr` argument has to be an rvalue anyway if it's a by-value parameter.

Reply:
changed